File: factory_Method.py
from abc import ABC, abstractmethod

class Vehiculo(ABC):
    @abstractmethod
    def conducir(self):
        pass

class Carro(Vehiculo):
    def conducir(self):
        return "Conduciendo un carro"

class Moto(Vehiculo):
    def conducir(self):
        return "Conduciendo una moto"

class Camion(Vehiculo):
    def conducir(self):
        return "Conduciendo un camión"

class FabricaVehiculo:
    def crear_vehiculo(self, tipo):
        if tipo == "carro":
            return Carro()
        elif tipo == "moto":
            return Moto()
        elif tipo == "camion":
            return Camion()
        else:
            raise ValueError("Tipo de vehículo no reconocido")


fabrica = FabricaVehiculo()
vehiculo = fabrica.crear_vehiculo("moto")
print(vehiculo.conducir())

fabrica2 = FabricaVehiculo()
vehiculo2 = fabrica2.crear_vehiculo("camion")
print(vehiculo2.conducir())

File: java/factory_method.java
abstract class Vehiculo {
    public abstract String conducir();
}

class Carro extends Vehiculo {
    @Override
    public String conducir() {
        return "Conduciendo un carro";
    }
}

class Moto extends Vehiculo {
    @Override
    public String conducir() {
        return "Conduciendo una moto";
    }
}

class Camion extends Vehiculo {
    @Override
    public String conducir() {
        return "Conduciendo un camión";
    }
}


class FabricaVehiculo {
    public Vehiculo crearVehiculo(String tipo) {
        switch (tipo.toLowerCase()) {
            case "carro":
                return new Carro();
            case "moto":
                return new Moto();
            case "camion":
                return new Camion();
            default:
                throw new IllegalArgumentException("Tipo de vehículo no reconocido");
        }
    }
}


public class factory_method {
    public static void main(String[] args) {
        FabricaVehiculo fabrica = new FabricaVehiculo();

        Vehiculo vehiculo1 = fabrica.crearVehiculo("moto");
        System.out.println(vehiculo1.conducir());

        Vehiculo vehiculo2 = fabrica.crearVehiculo("camion");
        System.out.println(vehiculo2.conducir());
    }
}
